File: Leetcode/Python3/783. Minimum Distance Between BST Nodes/solution.py
# https://leetcode.com/problems/minimum-distance-between-bst-nodes/


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    prev = float("inf")
    min_diff = float("inf")
    # Min_diff node will root.right node or rightmost node root.left
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        if root.left:
            self.minDiffInBST(root.left)
        
        self.min_diff = min(self.min_diff, abs(root.val - self.prev))
        self.prev = root.val

        if root.right:
            self.minDiffInBST(root.right)
        
        return self.min_diff


            

File: Leetcode/C++/783. Minimum Distance Between BST Nodes/solution.cpp
// https://leetcode.com/problems/minimum-distance-between-bst-nodes/

#include <vector>
#include <climits>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int minDiffInBST(TreeNode *root)
    {
        int diff = INT_MAX, prev = INT_MIN;
        vector<TreeNode *> stack;
        TreeNode *node = root;
        while (!stack.empty() || node)
        {
            // Move deepest left node
            while (node)
            {
                stack.push_back(node);
                node = node->left;
            }
            node = stack.back();
            stack.pop_back();
            // If current node not first search node check diff
            if (prev != INT_MIN)
            {
                diff = min(diff, node->val - prev);
            }
            prev = node->val;

            node = node->right;
        }
        return diff;
    }
};